lib/ts/controllers/s-modal.ts

```diff
@@ -0,0 +1,222 @@
+                if (e.target == lastTabbable && e.keyCode === 9 && !e.shiftKey) {
+                    e.preventDefault();
+                    firstTabbable.focus();
+                }
+            });
+
+            document.addEventListener("keydown", this._boundTabTrap);
+        }
+
+        /**
+         * Binds global events to the document for hiding popovers on user interaction
+         */
+        _bindDocumentEvents () {
+            // in order for removeEventListener to remove the right event, this bound function needs a constant reference
+            this._boundClickFn = this._boundClickFn || this._hideOnOutsideClick.bind(this);
+            this._boundKeypressFn = this._boundKeypressFn || this._hideOnEscapePress.bind(this);
+
+            document.addEventListener("click", this._boundClickFn);
+            document.addEventListener("keyup", this._boundKeypressFn);
+
+            this._bindTabFocusTrap();
+        }
+
+        /**
+         * Unbinds global events to the document for hiding popovers on user interaction
+         */
+        _unbindDocumentEvents () {
+            document.removeEventListener("click", this._boundClickFn);
+            document.removeEventListener("keyup", this._boundKeypressFn);
+            document.removeEventListener("keydown", this._boundTabTrap);
+        }
+
+        /**
+         * Forces the popover to hide if a user clicks outside of it or its reference element
+         */
+        _hideOnOutsideClick (e: Event) {
+            var target = <Node>e.target;
+            // check if the document was clicked inside either the toggle element or the modal itself
+            // note: .contains also returns true if the node itself matches the target element
+            if (!this.modalTarget.querySelector(".s-modal--dialog")!.contains(target)) {
+                this._toggle(false);
+            }
+        }
+
+        /**
+         * Forces the popover to hide if the user presses escape while it, one of its childen, or the reference element are focused
+         */
+        _hideOnEscapePress (e: KeyboardEvent) {
+            // if the ESC key (27) wasn't pressed or if no popovers are showing, return
+            if (e.which !== 27 || this.modalTarget.getAttribute("aria-hidden") === "true") {
+                return;
+            }
+
+            this._toggle(false);
+        }
+    }
+
+    /**
+     * Helper to manually show an s-modal element via external JS
+     * @param element the element the `data-controller="s-modal"` attribute is on
+     * @param show whether to force show/hide the modal; toggles the modal if left undefined
+     */
+    export function toggleModal(element: HTMLElement, show?: boolean | undefined) {
+        var controller = Stacks.application.getControllerForElementAndIdentifier(element, "s-modal") as ModalController;
+
+        if (!controller) {
+            throw "Unable to get s-modal controller from element";
```

Comment:
We should follow a similar pattern to s-tooltip.  If the controller is null and we're showing, `setAttribute("data-s-modal-show-on-connect", "true")`.  Then check and clear that data in `connect` and show the modal.

This is useful for circumstances where a modal was fetched from the server or created from a template.

Reply:
Punting for now. Will add in my next round of changes.